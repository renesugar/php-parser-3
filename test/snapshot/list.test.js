const parser = require('../main');

describe("Test list expressions", function() {
  it("test list statements", function() {
    const ast = parser.parseEval(`list($a => list($c,$d,,$e,), $b) = [1, 2];`, {
      ast: {
        withPositions: true
      }
    });
    expect(ast).toMatchSnapshot();
  });

  it("test short list", function() {
    expect(parser.parseEval(`[$a => [$c,$d,,$e,], $b] = [1 => [5, 6, 7, 8, 9,], 2];`, {
      ast: {
        withPositions: true
      }
    })).toMatchSnapshot();
  });

  it("fix #150", function() {
    const ast = parser.parseEval("list('foo' => $bar) = ['foo' => 'bar'];");
    expect(ast).toMatchSnapshot();
  });

  it("fix #137", function() {
    const ast = parser.parseEval("list(,,$a,,$b) = [null, 1, null, 2];", {
      ast: {
        withPositions: true
      }
    });
    expect(ast).toMatchSnapshot();
  });
});