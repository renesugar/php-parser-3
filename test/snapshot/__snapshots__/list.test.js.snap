// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test list expressions fix #137 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": Assign {
        "kind": "assign",
        "left": List {
          "items": Array [
            null,
            null,
            Variable {
              "byref": false,
              "curly": false,
              "kind": "variable",
              "loc": Location {
                "end": Position {
                  "column": 9,
                  "line": 1,
                  "offset": 9,
                },
                "source": null,
                "start": Position {
                  "column": 7,
                  "line": 1,
                  "offset": 7,
                },
              },
              "name": "a",
            },
            null,
            Variable {
              "byref": false,
              "curly": false,
              "kind": "variable",
              "loc": Location {
                "end": Position {
                  "column": 13,
                  "line": 1,
                  "offset": 13,
                },
                "source": null,
                "start": Position {
                  "column": 11,
                  "line": 1,
                  "offset": 11,
                },
              },
              "name": "b",
            },
          ],
          "kind": "list",
          "loc": Location {
            "end": Position {
              "column": 14,
              "line": 1,
              "offset": 14,
            },
            "source": null,
            "start": Position {
              "column": 0,
              "line": 1,
              "offset": 0,
            },
          },
          "shortForm": false,
        },
        "loc": Location {
          "end": Position {
            "column": 36,
            "line": 1,
            "offset": 36,
          },
          "source": null,
          "start": Position {
            "column": 0,
            "line": 1,
            "offset": 0,
          },
        },
        "operator": "=",
        "right": Array {
          "items": Array [
            Identifier {
              "kind": "identifier",
              "loc": Location {
                "end": Position {
                  "column": 22,
                  "line": 1,
                  "offset": 22,
                },
                "source": null,
                "start": Position {
                  "column": 18,
                  "line": 1,
                  "offset": 18,
                },
              },
              "name": ClassReference {
                "kind": "classreference",
                "loc": Location {
                  "end": Position {
                    "column": 22,
                    "line": 1,
                    "offset": 22,
                  },
                  "source": null,
                  "start": Position {
                    "column": 18,
                    "line": 1,
                    "offset": 18,
                  },
                },
                "name": "null",
                "resolution": "uqn",
              },
            },
            Number {
              "kind": "number",
              "loc": Location {
                "end": Position {
                  "column": 25,
                  "line": 1,
                  "offset": 25,
                },
                "source": null,
                "start": Position {
                  "column": 24,
                  "line": 1,
                  "offset": 24,
                },
              },
              "value": "1",
            },
            Identifier {
              "kind": "identifier",
              "loc": Location {
                "end": Position {
                  "column": 31,
                  "line": 1,
                  "offset": 31,
                },
                "source": null,
                "start": Position {
                  "column": 27,
                  "line": 1,
                  "offset": 27,
                },
              },
              "name": ClassReference {
                "kind": "classreference",
                "loc": Location {
                  "end": Position {
                    "column": 31,
                    "line": 1,
                    "offset": 31,
                  },
                  "source": null,
                  "start": Position {
                    "column": 27,
                    "line": 1,
                    "offset": 27,
                  },
                },
                "name": "null",
                "resolution": "uqn",
              },
            },
            Number {
              "kind": "number",
              "loc": Location {
                "end": Position {
                  "column": 34,
                  "line": 1,
                  "offset": 34,
                },
                "source": null,
                "start": Position {
                  "column": 33,
                  "line": 1,
                  "offset": 33,
                },
              },
              "value": "2",
            },
          ],
          "kind": "array",
          "loc": Location {
            "end": Position {
              "column": 35,
              "line": 1,
              "offset": 35,
            },
            "source": null,
            "start": Position {
              "column": 17,
              "line": 1,
              "offset": 17,
            },
          },
          "shortForm": true,
        },
      },
      "kind": "expressionstatement",
      "loc": Location {
        "end": Position {
          "column": 36,
          "line": 1,
          "offset": 36,
        },
        "source": null,
        "start": Position {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
    },
  ],
  "errors": Array [],
  "kind": "program",
  "loc": Location {
    "end": Position {
      "column": 36,
      "line": 1,
      "offset": 36,
    },
    "source": null,
    "start": Position {
      "column": 0,
      "line": 1,
      "offset": 0,
    },
  },
}
`;

exports[`Test list expressions fix #150 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": Assign {
        "kind": "assign",
        "left": List {
          "items": Array [
            Entry {
              "key": String {
                "isDoubleQuote": false,
                "kind": "string",
                "raw": "'foo'",
                "unicode": false,
                "value": "foo",
              },
              "kind": "entry",
              "value": Variable {
                "byref": false,
                "curly": false,
                "kind": "variable",
                "name": "bar",
              },
            },
          ],
          "kind": "list",
          "shortForm": false,
        },
        "operator": "=",
        "right": Array {
          "items": Array [
            Entry {
              "key": String {
                "isDoubleQuote": false,
                "kind": "string",
                "raw": "'foo'",
                "unicode": false,
                "value": "foo",
              },
              "kind": "entry",
              "value": String {
                "isDoubleQuote": false,
                "kind": "string",
                "raw": "'bar'",
                "unicode": false,
                "value": "bar",
              },
            },
          ],
          "kind": "array",
          "shortForm": true,
        },
      },
      "kind": "expressionstatement",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`Test list expressions test list statements 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": Assign {
        "kind": "assign",
        "left": List {
          "items": Array [
            Entry {
              "key": Variable {
                "byref": false,
                "curly": false,
                "kind": "variable",
                "loc": Location {
                  "end": Position {
                    "column": 7,
                    "line": 1,
                    "offset": 7,
                  },
                  "source": null,
                  "start": Position {
                    "column": 5,
                    "line": 1,
                    "offset": 5,
                  },
                },
                "name": "a",
              },
              "kind": "entry",
              "loc": Location {
                "end": Position {
                  "column": 27,
                  "line": 1,
                  "offset": 27,
                },
                "source": null,
                "start": Position {
                  "column": 5,
                  "line": 1,
                  "offset": 5,
                },
              },
              "value": List {
                "items": Array [
                  Variable {
                    "byref": false,
                    "curly": false,
                    "kind": "variable",
                    "loc": Location {
                      "end": Position {
                        "column": 18,
                        "line": 1,
                        "offset": 18,
                      },
                      "source": null,
                      "start": Position {
                        "column": 16,
                        "line": 1,
                        "offset": 16,
                      },
                    },
                    "name": "c",
                  },
                  Variable {
                    "byref": false,
                    "curly": false,
                    "kind": "variable",
                    "loc": Location {
                      "end": Position {
                        "column": 21,
                        "line": 1,
                        "offset": 21,
                      },
                      "source": null,
                      "start": Position {
                        "column": 19,
                        "line": 1,
                        "offset": 19,
                      },
                    },
                    "name": "d",
                  },
                  null,
                  Variable {
                    "byref": false,
                    "curly": false,
                    "kind": "variable",
                    "loc": Location {
                      "end": Position {
                        "column": 25,
                        "line": 1,
                        "offset": 25,
                      },
                      "source": null,
                      "start": Position {
                        "column": 23,
                        "line": 1,
                        "offset": 23,
                      },
                    },
                    "name": "e",
                  },
                  null,
                ],
                "kind": "list",
                "loc": Location {
                  "end": Position {
                    "column": 27,
                    "line": 1,
                    "offset": 27,
                  },
                  "source": null,
                  "start": Position {
                    "column": 11,
                    "line": 1,
                    "offset": 11,
                  },
                },
                "shortForm": false,
              },
            },
            Variable {
              "byref": false,
              "curly": false,
              "kind": "variable",
              "loc": Location {
                "end": Position {
                  "column": 31,
                  "line": 1,
                  "offset": 31,
                },
                "source": null,
                "start": Position {
                  "column": 29,
                  "line": 1,
                  "offset": 29,
                },
              },
              "name": "b",
            },
          ],
          "kind": "list",
          "loc": Location {
            "end": Position {
              "column": 32,
              "line": 1,
              "offset": 32,
            },
            "source": null,
            "start": Position {
              "column": 0,
              "line": 1,
              "offset": 0,
            },
          },
          "shortForm": false,
        },
        "loc": Location {
          "end": Position {
            "column": 42,
            "line": 1,
            "offset": 42,
          },
          "source": null,
          "start": Position {
            "column": 0,
            "line": 1,
            "offset": 0,
          },
        },
        "operator": "=",
        "right": Array {
          "items": Array [
            Number {
              "kind": "number",
              "loc": Location {
                "end": Position {
                  "column": 37,
                  "line": 1,
                  "offset": 37,
                },
                "source": null,
                "start": Position {
                  "column": 36,
                  "line": 1,
                  "offset": 36,
                },
              },
              "value": "1",
            },
            Number {
              "kind": "number",
              "loc": Location {
                "end": Position {
                  "column": 40,
                  "line": 1,
                  "offset": 40,
                },
                "source": null,
                "start": Position {
                  "column": 39,
                  "line": 1,
                  "offset": 39,
                },
              },
              "value": "2",
            },
          ],
          "kind": "array",
          "loc": Location {
            "end": Position {
              "column": 41,
              "line": 1,
              "offset": 41,
            },
            "source": null,
            "start": Position {
              "column": 35,
              "line": 1,
              "offset": 35,
            },
          },
          "shortForm": true,
        },
      },
      "kind": "expressionstatement",
      "loc": Location {
        "end": Position {
          "column": 42,
          "line": 1,
          "offset": 42,
        },
        "source": null,
        "start": Position {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
    },
  ],
  "errors": Array [],
  "kind": "program",
  "loc": Location {
    "end": Position {
      "column": 42,
      "line": 1,
      "offset": 42,
    },
    "source": null,
    "start": Position {
      "column": 0,
      "line": 1,
      "offset": 0,
    },
  },
}
`;

exports[`Test list expressions test short list 1`] = `
Program {
  "children": Array [
    ExpressionStatement {
      "expression": Assign {
        "kind": "assign",
        "left": List {
          "items": Array [
            Entry {
              "key": Variable {
                "byref": false,
                "curly": false,
                "kind": "variable",
                "loc": Location {
                  "end": Position {
                    "column": 3,
                    "line": 1,
                    "offset": 3,
                  },
                  "source": null,
                  "start": Position {
                    "column": 1,
                    "line": 1,
                    "offset": 1,
                  },
                },
                "name": "a",
              },
              "kind": "entry",
              "loc": Location {
                "end": Position {
                  "column": 19,
                  "line": 1,
                  "offset": 19,
                },
                "source": null,
                "start": Position {
                  "column": 1,
                  "line": 1,
                  "offset": 1,
                },
              },
              "value": Array {
                "items": Array [
                  Variable {
                    "byref": false,
                    "curly": false,
                    "kind": "variable",
                    "loc": Location {
                      "end": Position {
                        "column": 10,
                        "line": 1,
                        "offset": 10,
                      },
                      "source": null,
                      "start": Position {
                        "column": 8,
                        "line": 1,
                        "offset": 8,
                      },
                    },
                    "name": "c",
                  },
                  Variable {
                    "byref": false,
                    "curly": false,
                    "kind": "variable",
                    "loc": Location {
                      "end": Position {
                        "column": 13,
                        "line": 1,
                        "offset": 13,
                      },
                      "source": null,
                      "start": Position {
                        "column": 11,
                        "line": 1,
                        "offset": 11,
                      },
                    },
                    "name": "d",
                  },
                  null,
                  Variable {
                    "byref": false,
                    "curly": false,
                    "kind": "variable",
                    "loc": Location {
                      "end": Position {
                        "column": 17,
                        "line": 1,
                        "offset": 17,
                      },
                      "source": null,
                      "start": Position {
                        "column": 15,
                        "line": 1,
                        "offset": 15,
                      },
                    },
                    "name": "e",
                  },
                  null,
                ],
                "kind": "array",
                "loc": Location {
                  "end": Position {
                    "column": 19,
                    "line": 1,
                    "offset": 19,
                  },
                  "source": null,
                  "start": Position {
                    "column": 7,
                    "line": 1,
                    "offset": 7,
                  },
                },
                "shortForm": true,
              },
            },
            Variable {
              "byref": false,
              "curly": false,
              "kind": "variable",
              "loc": Location {
                "end": Position {
                  "column": 23,
                  "line": 1,
                  "offset": 23,
                },
                "source": null,
                "start": Position {
                  "column": 21,
                  "line": 1,
                  "offset": 21,
                },
              },
              "name": "b",
            },
          ],
          "kind": "list",
          "loc": Location {
            "end": Position {
              "column": 24,
              "line": 1,
              "offset": 24,
            },
            "source": null,
            "start": Position {
              "column": 0,
              "line": 1,
              "offset": 0,
            },
          },
          "shortForm": true,
        },
        "loc": Location {
          "end": Position {
            "column": 54,
            "line": 1,
            "offset": 54,
          },
          "source": null,
          "start": Position {
            "column": 0,
            "line": 1,
            "offset": 0,
          },
        },
        "operator": "=",
        "right": Array {
          "items": Array [
            Entry {
              "key": Number {
                "kind": "number",
                "loc": Location {
                  "end": Position {
                    "column": 29,
                    "line": 1,
                    "offset": 29,
                  },
                  "source": null,
                  "start": Position {
                    "column": 28,
                    "line": 1,
                    "offset": 28,
                  },
                },
                "value": "1",
              },
              "kind": "entry",
              "loc": Location {
                "end": Position {
                  "column": 49,
                  "line": 1,
                  "offset": 49,
                },
                "source": null,
                "start": Position {
                  "column": 28,
                  "line": 1,
                  "offset": 28,
                },
              },
              "value": Array {
                "items": Array [
                  Number {
                    "kind": "number",
                    "loc": Location {
                      "end": Position {
                        "column": 35,
                        "line": 1,
                        "offset": 35,
                      },
                      "source": null,
                      "start": Position {
                        "column": 34,
                        "line": 1,
                        "offset": 34,
                      },
                    },
                    "value": "5",
                  },
                  Number {
                    "kind": "number",
                    "loc": Location {
                      "end": Position {
                        "column": 38,
                        "line": 1,
                        "offset": 38,
                      },
                      "source": null,
                      "start": Position {
                        "column": 37,
                        "line": 1,
                        "offset": 37,
                      },
                    },
                    "value": "6",
                  },
                  Number {
                    "kind": "number",
                    "loc": Location {
                      "end": Position {
                        "column": 41,
                        "line": 1,
                        "offset": 41,
                      },
                      "source": null,
                      "start": Position {
                        "column": 40,
                        "line": 1,
                        "offset": 40,
                      },
                    },
                    "value": "7",
                  },
                  Number {
                    "kind": "number",
                    "loc": Location {
                      "end": Position {
                        "column": 44,
                        "line": 1,
                        "offset": 44,
                      },
                      "source": null,
                      "start": Position {
                        "column": 43,
                        "line": 1,
                        "offset": 43,
                      },
                    },
                    "value": "8",
                  },
                  Number {
                    "kind": "number",
                    "loc": Location {
                      "end": Position {
                        "column": 47,
                        "line": 1,
                        "offset": 47,
                      },
                      "source": null,
                      "start": Position {
                        "column": 46,
                        "line": 1,
                        "offset": 46,
                      },
                    },
                    "value": "9",
                  },
                  null,
                ],
                "kind": "array",
                "loc": Location {
                  "end": Position {
                    "column": 49,
                    "line": 1,
                    "offset": 49,
                  },
                  "source": null,
                  "start": Position {
                    "column": 33,
                    "line": 1,
                    "offset": 33,
                  },
                },
                "shortForm": true,
              },
            },
            Number {
              "kind": "number",
              "loc": Location {
                "end": Position {
                  "column": 52,
                  "line": 1,
                  "offset": 52,
                },
                "source": null,
                "start": Position {
                  "column": 51,
                  "line": 1,
                  "offset": 51,
                },
              },
              "value": "2",
            },
          ],
          "kind": "array",
          "loc": Location {
            "end": Position {
              "column": 53,
              "line": 1,
              "offset": 53,
            },
            "source": null,
            "start": Position {
              "column": 27,
              "line": 1,
              "offset": 27,
            },
          },
          "shortForm": true,
        },
      },
      "kind": "expressionstatement",
      "loc": Location {
        "end": Position {
          "column": 54,
          "line": 1,
          "offset": 54,
        },
        "source": null,
        "start": Position {
          "column": 0,
          "line": 1,
          "offset": 0,
        },
      },
    },
  ],
  "errors": Array [],
  "kind": "program",
  "loc": Location {
    "end": Position {
      "column": 54,
      "line": 1,
      "offset": 54,
    },
    "source": null,
    "start": Position {
      "column": 0,
      "line": 1,
      "offset": 0,
    },
  },
}
`;
